src/camera.py

```diff
@@ -0,0 +1,152 @@
+        """Extracts the camera ID from the URL."""
+        return "".join(char for char in self.url if char.isdigit())
+
+    def _get_camera_page(self, request_headers: Dict[str, str]) -> Optional[bytes]:
+        """Fetches the camera page content."""
+        try:
+            r = requests.get(self.url, headers=request_headers)
+            return r.content
+        except (RequestException, OSError) as e:
+            logger.error("error capturing image: " + str(e))
+            return None
+
+    def _find_camera_url(self) -> Optional[str]:
+        """Finds the camera stream URL."""
+        if self.page_tree is not None:
+            elems = self.page_tree.xpath("//img/@src")
+        if len(elems) > 0:
+            return elems[0]
+        return None
+
+    def _url_is_valid(self) -> bool:
+        """Checks if the stream URL is valid."""
+        if self.stream_url is None:
+            return False
+
+        return all(
+            [
+                self.stream_url != "/static/no.jpg",
+                "?stream" not in self.stream_url,
+                ".jpg" in self.stream_url,
+            ]
+        )
+
+    def _get_camera_details(self) -> Optional[str]:
+        """Extracts the camera details."""
+        details = self.page_tree.xpath('//div[@class="camera-details"]')
+        details_tuple = tuple(detail.text_content() for detail in details)
+        details = "".join(detail.replace("\n", "").replace("\t", "").strip() for detail in details_tuple)
+        return details
+
+    def _parse_camera_details(self) -> Optional[Dict[str, str]]:
+        """Parses the camera details and returns the camera info as a dictionary."""
+
+        if not self.details:
+            return None
+
+        details = self.details
+
+        city_start = details.find("City: ") + len("City: ")
+        city_end = details.find("Latitude:")
+        city = details[city_start:city_end]
+
+        region_start = details.find("Region:") + len("Region:")
+        region_end = details.find("City:")
+        region = details[region_start:region_end]
+
+        country_start = details.find("Country:") + len("Country:")
+        country_end = details.find("Country code:")
+        country = details[country_start:country_end]
+
+        country_code_start = details.find("Country code:") + len("Country code:")
+        country_code_end = details.find("Region:")
+        country_code = details[country_code_start:country_code_end]
+
+        return {
+            "city": city,
+            "region": region,
+            "country": country,
+            "country_code": country_code,
+        }
+
+    @exponential_backoff(
+        attempts=s.CAMERA_FETCH_ATTEMPTS,
+        initial_delay=s.INITIAL_RETRY_DELAY,
+        factor=s.RETRY_DELAY_FACTOR,
+        exception_types=(SaveImageException, FetchCamerasException),
+    )
+    def _save_image(self, image_file_path: str, camera_url: str, request_headers: Dict[str, str]) -> bool:

```

Comment:
Very impressed you got the back-off working here bud. (I haven't tested it but assume it does from the code...)

Did you find it particularly challenging?

Reply:
Thanks! Working on exponential backoff was a bit challenging, but not too tough. The main thing was getting a grip on decorators and how to use them correctly. Once I got that part down, putting together the backoff code was pretty smooth.

The decorator function wraps the original function and handles the backoff logic. If one of the specified exceptions pops up, it retries the function after waiting for the delay. The delay increases exponentially by multiplying it by the factor after each unsuccessful attempt. This keeps going until the function works or we hit the max attempts.